File: LeetCode/Java/src/main/java/org/redquark/tutorials/leetcode/ThreeSumClosest.java
package org.redquark.tutorials.leetcode;

import java.util.Arrays;

/**
 * @author Anirudh Sharma
 */
public class ThreeSumClosest {

    private static int threeSumClosest(int[] nums, int target) {
        // Sort the array
        Arrays.sort(nums);
        // Length of the array
        int n = nums.length;
        // Result
        int closest = nums[0] + nums[1] + nums[n - 1];
        // Loop for each element of the array
        for (int i = 0; i < n - 2; i++) {
            // Left and right pointers
            int j = i + 1;
            int k = n - 1;
            // Loop for all other pairs
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k];
                if (sum <= target) {
                    j++;
                } else {
                    k--;
                }
                if (Math.abs(closest - target) > Math.abs(sum - target)) {
                    closest = sum;
                }
            }
        }
        return closest;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{-1, 2, 1, -4};
        int target = 1;
        System.out.println(threeSumClosest(nums, target));
    }
}

File: LeetCode/Kotlin/src/main/kotlin/org/redquark/tutorials/leetcode/ThreeSumClosest.kt
package org.redquark.tutorials.leetcode

import java.util.*
import kotlin.math.abs

fun threeSumClosest(nums: IntArray, target: Int): Int {
    // Sort the array
    Arrays.sort(nums)
    // Length of the array
    val n = nums.size
    // Result
    var closest = nums[0] + nums[1] + nums[n - 1]
    // Loop for each element of the array
    for (i in 0 until n - 2) {
        // Left and right pointers
        var j = i + 1
        var k = n - 1
        // Loop for all other pairs
        while (j < k) {
            val sum = nums[i] + nums[j] + nums[k]
            if (sum <= target) {
                j++
            } else {
                k--
            }
            if (abs(closest - target) > abs(sum - target)) {
                closest = sum
            }
        }
    }
    return closest
}

fun main() {
    val nums = intArrayOf(-1, 2, 1, -4)
    val target = 1
    println(threeSumClosest(nums, target))
}